### ISSUE TITLE
Different results at each run

### ISSUE THREAD

**Issue by octoscore** [aksnzhy/xlearn#237]
Hi

I am not sure if we expect that but the following example predict different results each time I run it:
```
import numpy as np
import xlearn as xl
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split


iris_data = load_iris()
X = iris_data['data']
y = (iris_data['target'] == 2)

X_train,   \
X_val,     \
y_train,   \
y_val = train_test_split(X, y, test_size=0.3, random_state=0)

linear_model = xl.LRModel(task='binary', init=0.1,
                          epoch=10, lr=0.1,
                          reg_lambda=1.0, opt='sgd')


linear_model.fit(X_train, y_train,
                 eval_set=[X_val, y_val],
                 is_lock_free=False)


y_pred = linear_model.predict(X_val)
```

This is version 0.4.2 executed via sklearn api on windows: y_pred  is never the same. Are we expecting that?

Thank you

**Comment by etveritas** [comment]
@octoscore Hello, I test this snippet on my machine, and I find it has no change with several times. Could you give your results here, I just want to confirm whether the reason is lock free or something else.

**Comment by aksnzhy** [comment]
@octoscore The result of xLearn could be little different at each run because xLearn uses lock-free training. Typically it is doesn't manner for prediction. You can close lock-free training if you want to get a deterministic result. Thanks.

**Comment by octoscore** [comment]
This is really strange.  The results are so different each time. Are you testing on windows using Jupyter notebook? 

When I run the the same example in command line (example_LR_iris.py)  the test loss is: 0.740673 and it is not random at all.

Actually the first run in the notebook gives me the same answer but second run is different (without having killed the kernel)
first run : The test loss is: 0.740673
second run The test loss is: 0.067979

**Comment by etveritas** [comment]
@aksnzhy I found the same problem in jupyter on WSL. Does we have this problem before?

**Comment by aksnzhy** [comment]
@etveritas I didn't find this problem before.

**Comment by etveritas** [comment]
@aksnzhy I just test 0.4.1 version, and it also have different perform without resarting jupyter kernel.It seems that the model do not initialized unless restart the program.

**Comment by aksnzhy** [comment]
@etveritas Does this problem happen only in Windows?

**Comment by etveritas** [comment]
@aksnzhy  No, I also test this version on WSL(Windows Subsystem for Linux).

**Comment by etveritas** [comment]
@aksnzhy I run the ``mushroom`` demo seven times in jupyter, this demo use the native Python API, and this  is my run results about top 10 lines, the 1st to 6th without restart kernel, 7th time restart the kernel. Does this results in normal range of variation. I just want to check whether it's normal with native Python API firstly.(p.s. I build the master source code and run jupyter on windows.)
![image](https://user-images.githubusercontent.com/27916175/54610604-5008a280-4a90-11e9-8e97-1ee5483119f0.png)

**Comment by octoscore** [comment]
I think it is sensible. Is it lockfree=false? The differences I observed are much more different.

**Comment by octoscore** [comment]
Yes. Small change does not matter I am fine with it. But what we observed is really random.

The native api is fine for testing stuff but it is not enough for real life application. That is why we need the sklearn api.

**Comment by etveritas** [comment]
@octoscore Thanks for your advise, we will fix this problem as soon as possible. And we'd be happy if you could do more contribution to make this project be better. Thanks.

**Comment by etveritas** [comment]
@aksnzhy I get the reason why the results are different without restarting the jupyter(without lock-free). In jupyter the program will not end until we restart the kernel. Because we default [shuffle](https://github.com/aksnzhy/xlearn/blob/master/src/solver/solver.cc#L246-L248) for memory training, so xlearn will [generate different order of samples](https://github.com/aksnzhy/xlearn/blob/master/src/reader/reader.h#L203-L205) in one program life(i.e. without restarting jupyter kernel), then the results is different(without lock-free). So the results will be same(without lock-free) if we can keep the order genrating by ``random_shuffle`` is same in whole program life.

**Comment by aksnzhy** [comment]
@etveritas Thanks for this. I will try to fix this bug these days.

**Comment by etveritas** [comment]
@aksnzhy I create a PR, and I need your help to check for it.

**Comment by etveritas** [comment]
@octoscore Hello, we have fixed this bug, you can download and build the latest source code to avoid this problem.Thanks for your issue.

**Comment by octoscore** [comment]
Thank you very much. I will test asap. I hope to contribute more.

**Comment by etveritas** [comment]
@octoscore You can open issues or fork xlearn to add features what you think useful for users, we will resovle your issue or review your code. Thanks.
